contrib/sacrebleu/sacrebleu.py

```diff
@@ -0,0 +1,820 @@
+    :param rawfile: the input file (possibly SGML)
+    :param txtfile: the plaintext file
+    """
+
+    if not os.path.exists(txtfile):
+        if rawfile.endswith('.sgm') or rawfile.endswith('.sgml'):
+            logging.info("Processing {} to {}".format(rawfile, txtfile))
+            with _read(rawfile) as fin, open(txtfile, 'wt') as fout:
+                for line in fin:
+                    if line.startswith('<seg '):
+                        fout.write(re.sub(r'<seg.*?>(.*)</seg>.*?', '\\1', line))
+
+
+def print_test_set(test_set, langpair, side):
+    """Prints to STDOUT the specified side of the specified test set
+    :param test_set: the test set to print
+    :param langpair: the language pair
+    :param side: 'src' for source, 'ref' for reference
+    """
+
+    where = download_test_set(test_set, langpair)
+    infile = where[0] if side == 'src' else where[1]
+    with open(infile) as fin:
+        for line in fin:
+            print(line.rstrip())
+
+
+def download_test_set(test_set, langpair=None):
+    """Downloads the specified test to the system location specified by the SACREBLEU environment variable.
+    :param test_set: the test set to download
+    :param langpair: the language pair (needed for some datasets)
+    :return: the set of processed files
+    """
+
+    # if not data.has_key(test_set):
+    #     return None
+
+    dataset = data[test_set]['data']
+    outdir = os.path.join(SACREBLEU, test_set)
+    if not os.path.exists(outdir):
+        logging.info('Creating {}'.format(outdir))
+        os.makedirs(outdir)
+
+    tarball = os.path.join(outdir, os.path.basename(dataset))
+    rawdir = os.path.join(outdir, 'raw')
+    if not os.path.exists(tarball):
+        # TODO: check MD5sum
+        logging.info("Downloading {} to {}".format(dataset, tarball))
+        with urllib.request.urlopen(dataset) as f, open(tarball, 'wb') as out:
+            out.write(f.read())
+
+        # Extract the tarball
+        logging.info('Extracting {}'.format(tarball))
+        tar = tarfile.open(tarball)
+        tar.extractall(path=rawdir)
+
+    found = []
+
+    # Process the files into plain text
+    languages = data[test_set].keys() if langpair is None else [langpair]
+    for pair in languages:
+        if '-' not in pair:
+            continue
+        src, tgt = pair.split('-')
+        rawfile = os.path.join(rawdir, data[test_set][pair][0])
+        outfile = os.path.join(outdir, '{}.{}'.format(pair, src))
+        process_to_text(rawfile, outfile)
+        found.append(outfile)
+
+        rawfile = os.path.join(rawdir, data[test_set][pair][1])
+        outfile = os.path.join(outdir, '{}.{}'.format(pair, tgt))
+        process_to_text(rawfile, outfile)
+        found.append(outfile)
+
+    return found
+
+
+BLEU = namedtuple('BLEU', 'score, ngram1, ngram2, ngram3, ngram4, bp, sys_len, ref_len')
+
+
+def compute_bleu(instream, refstreams, smooth=0., force=False, lc=False, tokenize=False) -> BLEU:
```

Comment:
Without having any strong feelings about any of the BLEU flavours, are we going to sync this BLEU implementation with the one in sockeye? 

The latter has some minimal speed optimizations (for the WMT'17 bandit learning shared task):

```python
from random import choice
from string import ascii_uppercase
from collections import defaultdict
from itertools import  tee, islice
from sockeye.blue import zipngram

a=' '.join(choice(ascii_uppercase) for i in range(100))

%timeit extract_ngrams(a, 4)
10000 loops, best of 3: 129 µs per loop

%timeit for i in range(4): zipngram(a, i)
10000 loops, best of 3: 56.8 µs per loop
```

Reply:
Yes, if possible we should rely on sacrebleus BLEU implementation for the metric computation in sockeye.evaluate and the checkpoint_decoder.

Reply:
I tested this version and it's slower than the current one.

    def extract_ngrams_fast(line, max=4):
        tokens = line.split()
        ngrams = Counter()
        for n in range(1, max + 1):
            ngrams += Counter(zip(*(islice(it, pos, None) for pos, it in enumerate(tee(tokens, n + 1)))))
        return ngrams

    a = ' '.join(choice(ascii_uppercase) for i in range(100))
    print(timeit.timeit(lambda: extract_ngrams(a), number=10000))
    print(timeit.timeit(lambda: extract_ngrams_fast(a), number=10000))

This prints:

    2.814126971992664
    3.3099639220017707

But I will rejigger things so that Sockeye makes use of sacreBLEU and uses it as a requirement.